language/go/build_tags_test.go

```diff
@@ -0,0 +1,72 @@
+/* Copyright 2022 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package golang
+
+import (
+	"go/build/constraint"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestFilterBuildTags(t *testing.T) {
+	for _, tc := range []struct {
+		desc  string
+		input constraint.Expr
+		want  constraint.Expr
+	}{
+		{
+			desc:  "should remain",
+			input: mustParseBuildTag(t, "go1.8 || go1.9"),
+			want:  mustParseBuildTag(t, "go1.8 || go1.9"),
+		},
+		{
+			desc:  "simple 1",
+			input: mustParseBuildTag(t, "!(go1.8 || go1.9)"),
+			want:  mustParseBuildTag(t, "go1.8 && go1.9"),
+		},
+		{
+			desc:  "simple 2",
+			input: mustParseBuildTag(t, "!(foobar || go1.8 || go1.9)"),
+			want:  mustParseBuildTag(t, "!foobar && go1.8 && go1.9"),
+		},
+		{
+			desc:  "complex 1",
+			input: mustParseBuildTag(t, "!(cgo && (go1.8 || go1.9) || race || msan)"),
+			want:  mustParseBuildTag(t, "(cgo || (go1.8 && go1.9)) && race && msan"),
+		},
+		{
+			desc:  "complex 2",
+			input: mustParseBuildTag(t, "!(cgo && (go1.8 || go1.9 && (race && foobar)))"),
+			want:  mustParseBuildTag(t, "cgo || go1.8 && (go1.9 || (race || !foobar))"),
```

Comment:
ignored tags always evaluate to true (e.g. `!cgo == cgo == true`. This is the existing gazelle behavior. In order to replicate, we must first push negations down to the leaves of a boolean expression and then remove the negation for ignored tags. You can see here that because of this behavior, for test case `complex 2`, the user provided tag of `foobar` is never evaluated since the always true `cgo`, short circuits the boolean expression. I will leave it to y'all on how to feel about it.

Reply:
Unfortunately no. I took that approach first, but the new test cases I added (which caused the failures you pointed out) require that the ignored tags participate in the Boolean expression evaluation

Reply:
@sluongng What do you think?

Reply:
Gona put my thoughts in a separate review.